### PR TITLE
add xgboost v0.7.0

### PR DESCRIPTION
**Pull request by headupinclouds**

### CODE REVIEW

File: cmake/configs/default.cmake [hunk 1]
```diff
@@ -317,7 +317,7 @@ hunter_config(xext VERSION 1.3.1)
 hunter_config(xextproto VERSION 7.2.1)
 hunter_config(xf86vidmodeproto VERSION 2.3.1)
 hunter_config(xfixes VERSION 5.0.1)
-hunter_config(xgboost VERSION 0.40-p9)
+hunter_config(xgboost VERSION 0.7.0-p0)
```
Comment (ruslo) [inline cmake/configs/default.cmake]: 0.40 to 0.7? Is it downgrade?

Comment (headupinclouds) [reply cmake/configs/default.cmake]: I guess they have changed the versioning scheme to a strict semantic versioning:

https://github.com/dmlc/xgboost/issues/2925

> regularly publish minor updates, e.g. 0.7.1, 0.7.2, and so forth. Minor updates shall accommodate bug fixes and minor changes.

As you point out, the previous release used a slightly different versioning scheme:

https://github.com/dmlc/xgboost/releases

v0.40, v0.60

Comment (ruslo) [reply cmake/configs/default.cmake]: So in short 0.7 is newer than 0.40.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.3200
* https://travis-ci.org/ingenue/hunter/builds/326130796?utm_source=github_status&utm_medium=notification

File: cmake/configs/default.cmake (hunk 1 above)
Comment (headupinclouds) [reply cmake/configs/default.cmake]: Yes.

**Comment by headupinclouds** [comment]
Here is a brief summary of the various TOOLCHAIN errors.  I'll have to revisit these later.

Appveyor/Windows:  # MT vs MD flag mismatch
```
xgboost.lib(learner.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in foo.obj [C:\projects\hunter\_testing\_builds\vs-14-2015\foo.vcxproj]
```

linux + gcc-5: # Needs `-omp` flag I believe
```
(.text._ZN7xgboost9predictor12CPUPredictor19PredictContributionEPNS_7DMatrixEPSt6vectorIfSaIfEERKNS_3gbm11GBTreeModelEjb[_ZN7xgboost9predictor12CPUPredictor19PredictContributionEPNS_7DMatrixEPSt6vectorIfSaIfEERKNS_3gbm11GBTreeModelEjb]+0x35): undefined reference to `omp_get_max_threads'
cpu_predictor.cc:(.text._ZN7xgboost9predictor12CPUPredictor19PredictContributionEPNS_7DMatrixEPSt6vectorIfSaIfEERKNS_3gbm11GBTreeModelEjb[_ZN7xgboost9predictor12CPUPredictor19PredictContributionEPNS_7DMatrixEPSt6vectorIfSaIfEERKNS_3gbm11GBTreeModelEjb]+0x1ca): undefined reference to `GOMP_parallel'
cpu_predictor.cc:(.text._ZN7xgboost9predictor12CPUPredictor19PredictContributionEPNS_7DMatrixEPSt6vectorIfSaIfEERKNS_3gbm11GBTreeModelEjb[_ZN7xgboost9predictor12CPUPredictor19PredictContributionEPNS_7DMatrixEPSt6vectorIfSaIfEERKNS_3gbm11GBTreeModelEjb]+0x6ae): undefined reference to `GOMP_parallel'
```

linux + android-ndk-r15c-api-21-armeabi-v7a-neon-clang-libcxx: missing `#include`'s for android? 
```
/home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/../include/rabit/internal/utils.h:158:19: error: use of undeclared identifier 'fopen64'; did you mean 'fopen'?
  std::FILE *fp = fopen64(fname, flag);
                  ^~~~~~~
                  fopen
/home/travis/build/ingenue/hunter/_ci/android-ndk-r15c/sysroot/usr/include/stdio.h:237:7: note: 'fopen' declared here
FILE* fopen(const char* __restrict, const char* __restrict);
      ^
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/allreduce_base.cc:14:
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/./allreduce_base.h:18:
/home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/../include/rabit/internal/utils.h:158:19: error: use of undeclared identifier 'fopen64'; did you mean 'fopen'?
  std::FILE *fp = fopen64(fname, flag);
                  ^~~~~~~
                  fopen
/home/travis/build/ingenue/hunter/_ci/android-ndk-r15c/sysroot/usr/include/stdio.h:237:7: note: 'fopen' declared here
FILE* fopen(const char* __restrict, const char* __restrict);
      ^
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/allreduce_base.cc:14:
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/./allreduce_base.h:19:
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/../include/rabit/internal/engine.h:10:
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/src/../include/rabit/internal/../serializable.h:12:
In file included from /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/include/rabit/../dmlc/io.h:14:
/home/travis/build/ingenue/hunter/_testing/Hunter/_Base/7ea221c/97b2613/527b9c9/Build/rabit/Source/include/rabit/../dmlc/./logging.h:18:10: fatal error: 'execinfo.h' file not found
#include <execinfo.h>
```

**Comment by ruslo** [comment]
> MT vs MD flag mismatch

Blame:
* https://github.com/hunter-packages/xgboost/blob/edd22fda4ce8cc3d99cd10bc67b796b33c176a2e/cmake/Utils.cmake#L17-L33

Removing this line should fix it:
* https://github.com/hunter-packages/xgboost/blob/edd22fda4ce8cc3d99cd10bc67b796b33c176a2e/CMakeLists.txt#L14

> Needs -omp flag I believe

`target_compile_options(... PUBLIC ${OpenMP_CXX_FLAGS})` instead of modifying CMAKE_CXX_FLAGS:
* https://github.com/hunter-packages/xgboost/blob/edd22fda4ce8cc3d99cd10bc67b796b33c176a2e/CMakeLists.txt#L38-L40

> error: use of undeclared identifier 'fopen64'

From `stdio.h` on Android NDK:
```cpp
#if __ANDROID_API__ >= 24
FILE* fopen64(const char* __path, const char* __mode) __INTRODUCED_IN(24);
#endif /* __ANDROID_API__ >= 24 */
```

Currently API 21 used. If we need this I can add extra API 24 toolchain.

> fatal error: 'execinfo.h' file not found

There is no such file in Android NDK.

As far as I understand this functionality is optional:
* https://github.com/hunter-packages/rabit/blob/f08c6e0aa68a2f2fbf133307798cbfb17d6f688f/include/dmlc/logging.h#L17-L19

I think we should disable this feature on Android platform. Note that it's not xgboost, it's a rabit package.

**Comment by headupinclouds** [comment]
With gcc-5 this is also needed for proper linking:
```
    if(OpenMP_CXX_FOUND OR OPENMP_FOUND)    
      set_property(TARGET testxgboost APPEND PROPERTY LINK_FLAGS ${OpenMP_CXX_FLAGS})
    endif()
```

I've added the `-fopenmp` and `MSVC` fixes here: https://github.com/hunter-packages/xgboost/pull/13

> I think we should disable this feature on Android platform. Note that it's not xgboost, it's a rabit package.

I'll look into this one tomorrow.

**Comment by ruslo** [comment]
> With gcc-5 this is also needed for proper linking

I think `target_link_libraries` can be used here.

**Comment by headupinclouds** [comment]
It actually has a test, but for `__ANDROID__`

https://github.com/hunter-packages/dmlc-core/blob/048e90ef6d41f4ebfd05116e50639540c989368b/include/dmlc/base.h#L44-L48
```
#if (defined(__GNUC__) && !defined(__MINGW32__)\
     && !(defined __MINGW64__) && !(defined __ANDROID__))
#if (!defined(DMLC_LOG_STACK_TRACE))
#define DMLC_LOG_STACK_TRACE 1
#endif
```

**Comment by headupinclouds** [comment]
> I think target_link_libraries can be used here.

Thanks.  That's better.

**Comment by ruslo** [comment]
> It actually has a test, but for `__ANDROID__`
> https://github.com/hunter-packages/dmlc-core

Not in the **rabit** package:
* https://github.com/hunter-packages/rabit/blob/f08c6e0aa68a2f2fbf133307798cbfb17d6f688f/include/dmlc/base.h#L44-L46

Bundled dependencies are evil. Hunterization will do the trick.

**Comment by headupinclouds** [comment]
> Not in the rabit package:

I missed the bundled headers during packaging.  Tomorrow...

**Comment by headupinclouds** [comment]
Actually, I'm still seeing an openmp error with the hunter test using the installed xgboost in the android toolchain:

```
cd hunter
TOOLCHAIN=android-ndk-r16b-api-24-armeabi-v7a-neon-clang-libcxx PROJECT_DIR=examples/xgboost ./jenkins.py
```


Although the local build is working:
```
polly.py --toolchain android-ndk-r16b-api-24-armeabi-v7a-neon-clang-libcxx --config Release --clear --fwd BUILD_EXAMPLE=ON BUILD_TEST=ON
```

**Comment by ruslo** [comment]
All calls to `target_link_libraries(testxgboost PUBLIC ${OpenMP_CXX_FLAGS})` are under some conditions, which is `OFF`: `if(XGBOOST_BUILD_EXAMPLE)`, `if(GOOGLE_TEST)`.

**Comment by headupinclouds** [comment]
> All calls to target_link_libraries(testxgboost PUBLIC ${OpenMP_CXX_FLAGS})

That's  fixed in xgboost 0.7.0-2 https://github.com/ruslo/hunter/pull/1295/commits/da04653edaafda1dabb36a8ab4e0ffda7455adbba.  I've applied the linking to the library targets (not the executables) so they'll be propagated to the downstream consumers as well.  Also, since FindCUDA.cmake uses the "plain" `target_link_libraries`, we can't use `PUBLIC,PRIVATE`.

**Comment by ruslo** [comment]
> Also, since FindCUDA.cmake uses the "plain" target_link_libraries, we can't use PUBLIC,PRIVATE

Fix is to add `PUBLIC` everywhere. Actually Android seems to work on my local machine, so testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.3211
* https://travis-ci.org/ingenue/hunter/builds/326811996?utm_source=github_status&utm_medium=notification

File: cmake/configs/default.cmake [hunk 2]
```diff
@@ -175,7 +175,7 @@ hunter_config(damageproto VERSION 1.2.1)
 hunter_config(dbus VERSION 1.10.0-hunter-4)
 hunter_config(debug_assert VERSION 1.3)
 hunter_config(dest VERSION 0.8.0-p4)
-hunter_config(dlib VERSION 19.6-p2)
+hunter_config(dlib VERSION 19.8-p0)
```
Comment (ruslo) [inline cmake/configs/default.cmake]: Doesn't belong here, can you please rebase?

**Comment by headupinclouds** [comment]
> Fix is to add PUBLIC everywhere. Actually Android seems to work on my local machine, so testing:

It is using the plain signatures consistently now, so it will work.  I just realized `FindCUDA` supports `CUDA_LINK_LIBRARIES_KEYWORD` for this purpose:

https://github.com/Kitware/CMake/blob/89d12746306571a323549ad0f8d11284999053d5/Modules/FindCUDA.cmake#L1815-L1817

File: cmake/configs/default.cmake (hunk 2 above)
Comment (headupinclouds) [reply cmake/configs/default.cmake]: I'll take a look in > 1 hour.

**Comment by ruslo** [comment]
`/MT` / `/MD` conflict in dependency `dmlc`:
```
dmlc.lib(local_filesys.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in foo.obj [C:\projects\hunter\_testing\_builds\vs-14-2015\foo.vcxproj]
```

**Comment by headupinclouds** [comment]
Probably all of these compiler flags could be removed and enforced in toolchains as needed: 

https://github.com/hunter-packages/dmlc-core/blob/fdaad7fbd168864176c4003980bc0cdfdf69c64d/CMakeLists.txt#L74-L109

This seems to be a recurring anti-pattern in the context of package management.

**Comment by headupinclouds** [comment]
For now, I'll remove this:
https://github.com/hunter-packages/dmlc-core/blob/fdaad7fbd168864176c4003980bc0cdfdf69c64d/CMakeLists.txt#L76-L81

**Comment by ruslo** [comment]
> Probably all of these compiler flags could be removed and enforced in toolchains as needed

Most of them, everything except warnings and optimization. Don't know what is the point of touching something like `-g`, `-O0`, `-O3` in the first place.

> This seems to be a recurring anti-pattern in the context of package management

Yep. If you have a set of packages to compile and if at least one of them use `/MT` instead of default `/MD` you have broken build. Hence `/MT` should be set simultaneously and globally for all of them, e.g. in toolchain.

**Comment by headupinclouds** [comment]
I've updated `dmlc-core` here https://github.com/ruslo/hunter/pull/1300

**Comment by headupinclouds** [comment]
Note: Tests waiting for #1300

**Comment by headupinclouds** [comment]
This one includes the dmlc-core MSVC fixes and should be ready for testing.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.3225
* https://travis-ci.org/ingenue/hunter/builds/327575022?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Uploading:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.3228
* https://travis-ci.org/ingenue/hunter/builds/327616032?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.227

**Pull request merged**